### PR TITLE
Backend/Feature/#691/Added Updated Driver location and Updated Actual…

### PR DESCRIPTION
**Pull request by abhigyanshukla97**
… Distance to Pickup driver side and customer side

## Type of Change

- [ ] Bugfix
- [x] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates

## Description
 Added functionality for both drive side and customer side (table and changes).

### Additional Changes

- [x] This PR modifies the database schema (database migration added)
- [ ] This PR modifies dhall configs/environment variables
- [ ] This PR contains API breaking changes

## Checklist

- [ ] I formatted the code and addressed linter errors `./dev/format-all-files.sh`
- [ ] I reviewed submitted code
- [ ] I added unit tests for my changes where possible
- [ ] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable

### CODE REVIEW

**Review by roman-bodavskiy — CHANGES_REQUESTED** [review]

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/Ride/CancelRide.hs [hunk 1]
```diff
@@ -86,22 +94,25 @@ cancelRideHandler ServiceHandle {..} requestorId rideId req = withLogTag ("rideI
...
           unless (authPerson.merchantId == driver.merchantId) $ throwError (RideDoesNotExist rideId.getId)
           logTagInfo "admin -> cancelRide : " ("DriverId " <> getId driverId <> ", RideId " <> getId ride.id)
-          buildRideCancelationReason Nothing DBCR.ByMerchant ride
+          buildRideCancelationReason Nothing Nothing Nothing Nothing DBCR.ByMerchant ride
         DP.DRIVER -> do
           unless (authPerson.id == driverId) $ throwError NotAnExecutor
           logTagInfo "driver -> cancelRide : " ("DriverId " <> getId driverId <> ", RideId " <> getId ride.id)
-          buildRideCancelationReason (Just driverId) DBCR.ByDriver ride
+          location <- findDriverLocationId driverId >>= fromMaybeM LocationNotFound
+          loctopickup <- findDriverDistance driverId >>= fromMaybeM (InternalError $ "Distance Not Found")
```
[the remark this answers is not in the document]
Comment (roman-bodavskiy) [reply Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/Ride/CancelRide.hs:102]: My opinion , you can fetch actual distance to pickup from DB just wthe same as we fetch it when calculating driver pool :

```
        location ^. DriverLocationPoint <->. Esq.getPoint (val lat, val lon),
        location ^. DriverLocationLat,
        location ^. DriverLocationLon,
```

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/Ride/CancelRide.hs [hunk 2]
```diff
@@ -42,15 +46,19 @@ type MonadHandler m = (MonadThrow m, Log m, MonadGuid m)
 data ServiceHandle m = ServiceHandle
   { findRideById :: Id DRide.Ride -> m (Maybe DRide.Ride),
     findById :: Id DP.Person -> m (Maybe DP.Person),
-    cancelRide :: Id DRide.Ride -> DBCR.BookingCancellationReason -> m ()
+    findDriverLocationId :: Id DP.Person -> m (Maybe DDriverLocation.DriverLocation),
+    cancelRide :: Id DRide.Ride -> DBCR.BookingCancellationReason -> m (),
+    findDriverDistance :: Id DP.Person -> m (Maybe DDriverQuote.DriverQuote)
```
[the remark this answers is not in the document]
Comment (abhigyanshukla97) [reply Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/Ride/CancelRide.hs]: Pushed new changes.

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/Ride/CancelRide.hs [hunk 3]
```diff
@@ -86,22 +94,25 @@ cancelRideHandler ServiceHandle {..} requestorId rideId req = withLogTag ("rideI
...
-          buildRideCancelationReason Nothing DBCR.ByMerchant ride
+          buildRideCancelationReason Nothing Nothing Nothing Nothing DBCR.ByMerchant ride
         DP.DRIVER -> do
           unless (authPerson.id == driverId) $ throwError NotAnExecutor
           logTagInfo "driver -> cancelRide : " ("DriverId " <> getId driverId <> ", RideId " <> getId ride.id)
-          buildRideCancelationReason (Just driverId) DBCR.ByDriver ride
+          location <- findDriverLocationId driverId >>= fromMaybeM LocationNotFound
+          loctopickup <- findDriverDistance driverId >>= fromMaybeM (InternalError $ "Distance Not Found")
+          buildRideCancelationReason (Just location.lat) (Just location.lon) (Just loctopickup.distanceToPickup) (Just driverId) DBCR.ByDriver ride
+    --buildRideCancelationReason Nothing Nothing Nothing (Just driverId) DBCR.ByDriver ride
```
[the remark this answers is not in the document]
Comment (abhigyanshukla97) [reply Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/Ride/CancelRide.hs]: Pushed new changes.

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/Ride/CancelRide.hs (hunk 1 above)
[the remark this answers is not in the document]
Comment (abhigyanshukla97) [reply Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/Ride/CancelRide.hs]: Pushed new changes.

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/Beckn/Cancel.hs [hunk 4]
```diff
@@ -119,6 +119,9 @@ cancel req merchant booking = do
             reasonCode = Nothing,
             driverId = Nothing,
             additionalInfo = Nothing,
+            driverLat = Nothing,
```
Comment (roman-bodavskiy) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/Beckn/Cancel.hs:122]: why do we have `Nothing` here? Can we also use current location?

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/Ride/CancelRide.hs [hunk 5]
```diff
@@ -42,15 +48,21 @@ type MonadHandler m = (MonadThrow m, Log m, MonadGuid m)
...
 cancelRideHandle :: ServiceHandle Flow
 cancelRideHandle =
   ServiceHandle
     { findRideById = QRide.findById,
       findById = QPerson.findById,
-      cancelRide = CInternal.cancelRideImpl
+      cancelRide = CInternal.cancelRideImpl,
+      findDriverLocationId = QDrLoc.findById,
+      findBookingById = QRB.findById,
+      pickUpDistance = RideEndInt.getDistanceBetweenPoints
```
[the remark this answers is not in the document]
Comment (roman-bodavskiy) [reply Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/Ride/CancelRide.hs:65]: Also, this change require one more google api call. Is it acceptable for us?

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/Beckn/Cancel.hs (hunk 4 above)
Comment (abhigyanshukla97) [reply Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/Beckn/Cancel.hs:122]: Its on driver side when user cancels we need to add to the customer side thats why left Nothing.

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/Ride/CancelRide.hs [hunk 6]
```diff
@@ -86,22 +98,25 @@ cancelRideHandler ServiceHandle {..} requestorId rideId req = withLogTag ("rideI
...
           unless (authPerson.merchantId == driver.merchantId) $ throwError (RideDoesNotExist rideId.getId)
           logTagInfo "admin -> cancelRide : " ("DriverId " <> getId driverId <> ", RideId " <> getId ride.id)
-          buildRideCancelationReason Nothing DBCR.ByMerchant ride
+          buildRideCancelationReason Nothing Nothing Nothing Nothing DBCR.ByMerchant ride
         DP.DRIVER -> do
           unless (authPerson.id == driverId) $ throwError NotAnExecutor
           logTagInfo "driver -> cancelRide : " ("DriverId " <> getId driverId <> ", RideId " <> getId ride.id)
-          buildRideCancelationReason (Just driverId) DBCR.ByDriver ride
+          location <- findDriverLocationId driverId >>= fromMaybeM LocationNotFound
+          booked <- findBookingById ride.bookingId >>= fromMaybeM (BookingNotFound ride.bookingId.getId)
```
[the remark this answers is not in the document]
Comment (abhigyanshukla97) [reply Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/Ride/CancelRide.hs]: Renamed.

File: Backend/dev/migrations/rider-app/1117-add-column-to-merchant_service_usage_config.sql [hunk 7]
```diff
@@ -0,0 +1,2 @@
+ALTER TABLE atlas_app.merchant_service_usage_config ADD COLUMN get_distances_o_s_r_m text;
+UPDATE atlas_app.merchant_service_usage_config SET get_distances_o_s_r_m ='OSRM';
```
Comment (roman-bodavskiy) [inline Backend/dev/migrations/rider-app/1117-add-column-to-merchant_service_usage_config.sql:2]: Hm why did you add this field?
If it will be always OSRM, then no need to add this field
If it may be changed, then name is not proper, can be something like `get_distances_for_cancel_ride`

File: Backend/app/rider-platform/rider-app/Main/src/Storage/Tabular/Merchant/MerchantServiceUsageConfig.hs [hunk 8]
```diff
@@ -52,6 +52,7 @@ mkPersist
       whatsappProvidersPriorityList (PostgresList WhatsappService)
       updatedAt UTCTime
       createdAt UTCTime
+      getDistancesOSRM MapsService
```
Comment (roman-bodavskiy) [inline Backend/app/rider-platform/rider-app/Main/src/Storage/Tabular/Merchant/MerchantServiceUsageConfig.hs:55]: Can you move your field before `createdAt` and `updatedAt`. This two fields should be the last in all types

File: Backend/app/rider-platform/rider-app/Main/src/Domain/Types/Merchant/MerchantServiceUsageConfig.hs [hunk 9]
```diff
@@ -40,7 +40,8 @@ data MerchantServiceUsageConfigD (s :: UsageSafety) = MerchantServiceUsageConfig
     whatsappProvidersPriorityList :: [WhatsappService],
     useFraudDetection :: Bool,
     updatedAt :: UTCTime,
-    createdAt :: UTCTime
+    createdAt :: UTCTime,
+    getDistancesOSRM :: MapsService
```
Comment (roman-bodavskiy) [inline Backend/app/rider-platform/rider-app/Main/src/Domain/Types/Merchant/MerchantServiceUsageConfig.hs:44]: the same, can you move up?

File: Backend/dev/migrations/rider-app/1117-add-column-to-merchant_service_usage_config.sql (hunk 7 above)
Comment (abhigyanshukla97) [reply Backend/dev/migrations/rider-app/1117-add-column-to-merchant_service_usage_config.sql]: Pushed new changes.

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/Ride/CancelRide.hs [hunk 10]
```diff
@@ -86,22 +101,25 @@ cancelRideHandler ServiceHandle {..} requestorId rideId req = withLogTag ("rideI
...
           logTagInfo "admin -> cancelRide : " ("DriverId " <> getId driverId <> ", RideId " <> getId ride.id)
-          buildRideCancelationReason Nothing DBCR.ByMerchant ride
+          buildRideCancelationReason Nothing Nothing Nothing Nothing DBCR.ByMerchant ride
         DP.DRIVER -> do
           unless (authPerson.id == driverId) $ throwError NotAnExecutor
           logTagInfo "driver -> cancelRide : " ("DriverId " <> getId driverId <> ", RideId " <> getId ride.id)
-          buildRideCancelationReason (Just driverId) DBCR.ByDriver ride
+          location <- findDriverLocationId driverId >>= fromMaybeM LocationNotFound
+          booking <- findBookingById ride.bookingId >>= fromMaybeM (BookingNotFound ride.bookingId.getId)
+          locToPickup <- pickUpDistance booking.providerId (getCoordinates location) (getCoordinates booking.fromLocation)
```
Comment (piyushKumar-1) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/Ride/CancelRide.hs:110]: can you name is `disToPickup` instead of `locToPickup` ?

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/Ride/CancelRide.hs [hunk 11]
```diff
@@ -86,22 +101,25 @@ cancelRideHandler ServiceHandle {..} requestorId rideId req = withLogTag ("rideI
...
           unless (authPerson.merchantId == driver.merchantId) $ throwError (RideDoesNotExist rideId.getId)
           logTagInfo "admin -> cancelRide : " ("DriverId " <> getId driverId <> ", RideId " <> getId ride.id)
-          buildRideCancelationReason Nothing DBCR.ByMerchant ride
+          buildRideCancelationReason Nothing Nothing Nothing Nothing DBCR.ByMerchant ride
         DP.DRIVER -> do
           unless (authPerson.id == driverId) $ throwError NotAnExecutor
           logTagInfo "driver -> cancelRide : " ("DriverId " <> getId driverId <> ", RideId " <> getId ride.id)
-          buildRideCancelationReason (Just driverId) DBCR.ByDriver ride
+          location <- findDriverLocationId driverId >>= fromMaybeM LocationNotFound
+          booking <- findBookingById ride.bookingId >>= fromMaybeM (BookingNotFound ride.bookingId.getId)
```
Comment (piyushKumar-1) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/Ride/CancelRide.hs:109]: read this `findBookingById` from replica

File: Backend/app/rider-platform/rider-app/Main/src/Domain/Action/Beckn/OnUpdate.hs [hunk 12]
```diff
@@ -451,5 +451,8 @@ buildBookingCancellationReason bookingId mbRideId cancellationSource = do
         source = cancellationSource,
         reasonCode = Nothing,
         reasonStage = Nothing,
-        additionalInfo = Nothing
+        additionalInfo = Nothing,
+        driverLat = Nothing,
```
[the remark this answers is not in the document]
Comment (abhigyanshukla97) [reply Backend/app/rider-platform/rider-app/Main/src/Domain/Action/Beckn/OnUpdate.hs:455]: yes as its for analytics.

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Storage/Tabular/Merchant/MerchantServiceUsageConfig.hs [hunk 13]
```diff
@@ -50,6 +50,7 @@ mkPersist
       smsProvidersPriorityList (PostgresList SmsService)
       whatsappProvidersPriorityList (PostgresList WhatsappService)
       verificationService VerificationService
+      getDistancesForCancelRide MapsService
```
[the remark this answers is not in the document]
Comment (abhigyanshukla97) [reply Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Storage/Tabular/Merchant/MerchantServiceUsageConfig.hs:53]: yes i have made the changes.

File: Backend/app/rider-platform/rider-app/Main/src/Storage/Tabular/Merchant/MerchantServiceUsageConfig.hs [hunk 14]
```diff
@@ -50,6 +50,7 @@ mkPersist
       useFraudDetection Bool
       smsProvidersPriorityList (PostgresList SmsService)
       whatsappProvidersPriorityList (PostgresList WhatsappService)
+      getDistancesForCancelRide MapsService
```
[the remark this answers is not in the document]
Comment (abhigyanshukla97) [reply Backend/app/rider-platform/rider-app/Main/src/Storage/Tabular/Merchant/MerchantServiceUsageConfig.hs:53]: yes i have made the changes.

File: Backend/app/rider-platform/rider-app/Main/src/Domain/Action/UI/Cancel.hs [hunk 15]
```diff
@@ -82,7 +88,10 @@ cancel bookingId _ req = do
     throwError $ RideInvalidStatus "Cannot cancel this ride"
   when (booking.status == SRB.NEW) $ throwError (BookingInvalidStatus "NEW")
   bppBookingId <- fromMaybeM (BookingFieldNotPresent "bppBookingId") booking.bppBookingId
-  cancellationReason <- buildBookingCancelationReason
+  ride <- QR.findActiveByRBId booking.id >>= fromMaybeM (RideDoesNotExist $ "BookingId: " <> booking.id.getId)
```
[the remark this answers is not in the document]
Comment (abhigyanshukla97) [reply Backend/app/rider-platform/rider-app/Main/src/Domain/Action/UI/Cancel.hs:91]: yes i have made the change.

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/Ride/CancelRide.hs (hunk 10 above)
Comment (abhigyanshukla97) [reply Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/Ride/CancelRide.hs:110]: changed.

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/Ride/CancelRide.hs (hunk 11 above)
Comment (abhigyanshukla97) [reply Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/Ride/CancelRide.hs]: yes i have made the changes.

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/Dashboard/Booking.hs [hunk 16]
```diff
@@ -91,7 +91,10 @@ mkBookingCancellationReason reasonCode mbRideId bookingId = do
       source = DBCR.ByMerchant,
       reasonCode = Just $ coerce @Common.CancellationReasonCode @DCR.CancellationReasonCode reasonCode,
       driverId = Nothing,
-      additionalInfo = Nothing
+      additionalInfo = Nothing,
```
Comment (hkmangla) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/Dashboard/Booking.hs:94]: updateKeys.

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/Ride/CancelRide.hs [hunk 17]
```diff
@@ -86,22 +102,25 @@ cancelRideHandler ServiceHandle {..} requestorId rideId req = withLogTag ("rideI
           driver <- findById driverId >>= fromMaybeM (PersonNotFound driverId.getId)
           unless (authPerson.merchantId == driver.merchantId) $ throwError (RideDoesNotExist rideId.getId)
           logTagInfo "admin -> cancelRide : " ("DriverId " <> getId driverId <> ", RideId " <> getId ride.id)
-          buildRideCancelationReason Nothing DBCR.ByMerchant ride
+          buildRideCancelationReason Nothing Nothing Nothing Nothing DBCR.ByMerchant ride
         DP.DRIVER -> do
           unless (authPerson.id == driverId) $ throwError NotAnExecutor
           logTagInfo "driver -> cancelRide : " ("DriverId " <> getId driverId <> ", RideId " <> getId ride.id)
-          buildRideCancelationReason (Just driverId) DBCR.ByDriver ride
+          location <- findDriverLocationId driverId >>= fromMaybeM LocationNotFound
```
Comment (hkmangla) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/Ride/CancelRide.hs:109]: If driver location doesn't find, don't throw error, just update it as Nothing. As cancellation should not fail if we don't have driver location.

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Storage/Tabular/BookingCancellationReason.hs [hunk 18]
```diff
@@ -40,6 +41,9 @@ mkPersist
       source Domain.CancellationSource
       reasonCode CancellationReasonTId Maybe
       additionalInfo Text Maybe
+      driverLat Double Maybe
```
Comment (hkmangla) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Storage/Tabular/BookingCancellationReason.hs:44]: Here keys are find, only need to change in Domain as comment added above.

File: Backend/app/rider-platform/rider-app/Main/src/Domain/Action/Beckn/OnUpdate.hs (hunk 12 above)
[the remark this answers is not in the document]
Comment (hkmangla) [reply Backend/app/rider-platform/rider-app/Main/src/Domain/Action/Beckn/OnUpdate.hs:455]: Please fill in the data as suggested in 1st comment, as we need data at both side.

File: Backend/app/rider-platform/rider-app/Main/src/Domain/Action/Dashboard/Booking.hs [hunk 19]
```diff
@@ -72,7 +72,10 @@ mkBookingCancellationReason reasonCode mbRideId bookingId = do
       source = DBCR.ByMerchant,
       reasonCode = Just $ coerce @Common.CancellationReasonCode @DCR.CancellationReasonCode reasonCode,
       reasonStage = Nothing,
-      additionalInfo = Nothing
+      additionalInfo = Nothing,
+      driverLat = Nothing,
```
Comment (hkmangla) [inline Backend/app/rider-platform/rider-app/Main/src/Domain/Action/Dashboard/Booking.hs:76]: rename keys properly as suggested at BPP side.

File: Backend/app/rider-platform/rider-app/Main/src/Domain/Action/UI/Cancel.hs [hunk 20]
```diff
@@ -156,3 +168,25 @@ cancelSearch personId dcr = do
...
+    EsqDBFlow m r,
+    Metrics.CoreMetrics m,
+    Maps.HasCoordinates a,
+    Maps.HasCoordinates b
+  ) =>
+  Id Merchant ->
+  a ->
+  b ->
+  m Meters
+getDistanceBetweenPoints merchantId a b = do
```
Comment (hkmangla) [inline Backend/app/rider-platform/rider-app/Main/src/Domain/Action/UI/Cancel.hs:184]: don't use variable name as `a` and `b`. Rename them properly.

File: Backend/app/rider-platform/rider-app/Main/src/Domain/Action/UI/Cancel.hs [hunk 21]
```diff
@@ -82,7 +88,10 @@ cancel bookingId _ req = do
     throwError $ RideInvalidStatus "Cannot cancel this ride"
   when (booking.status == SRB.NEW) $ throwError (BookingInvalidStatus "NEW")
   bppBookingId <- fromMaybeM (BookingFieldNotPresent "bppBookingId") booking.bppBookingId
-  cancellationReason <- buildBookingCancelationReason
+  ride <- Esq.runInReplica $ QR.findActiveByRBId booking.id >>= fromMaybeM (RideDoesNotExist $ "BookingId: " <> booking.id.getId)
+  currLocation <- RLOC.getDriverLoc ride.id personId
```
Comment (hkmangla) [inline Backend/app/rider-platform/rider-app/Main/src/Domain/Action/UI/Cancel.hs:92]: Don't call Domain.UI function in other Domain file. Directly use CallBPP to get the driver location.

File: Backend/app/rider-platform/rider-app/Main/src/Domain/Action/UI/Cancel.hs [hunk 22]
```diff
@@ -82,7 +88,10 @@ cancel bookingId _ req = do
...
+  finalDis <- getDistanceBetweenPoints booking.merchantId (getCoordinates currLocation) (getCoordinates booking.fromLocation)
```
Comment (hkmangla) [inline Backend/app/rider-platform/rider-app/Main/src/Domain/Action/UI/Cancel.hs:93]: Rename variable from `finalDes` to something proper.

File: Backend/app/rider-platform/rider-app/Main/src/Domain/Action/UI/Cancel.hs [hunk 23]
```diff
@@ -156,3 +168,25 @@ cancelSearch personId dcr = do
         Esq.runTransaction $ QPFS.updateStatus personId DPFS.IDLE
         QEstimate.updateStatus dcr.estimateId DEstimate.CANCELLED
   QPFS.clearCache personId
+
+getDistanceBetweenPoints ::
```
Comment (hkmangla) [inline Backend/app/rider-platform/rider-app/Main/src/Domain/Action/UI/Cancel.hs:172]: Rename this function properly.

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/Ride/CancelRide.hs [hunk 24]
```diff
@@ -110,5 +129,30 @@ cancelRideHandler ServiceHandle {..} requestorId rideId req = withLogTag ("rideI
...
+    EsqDBFlow m r,
+    Metrics.CoreMetrics m,
+    Maps.HasCoordinates a,
+    Maps.HasCoordinates b
+  ) =>
+  Id Merchant ->
+  a ->
+  b ->
+  m Meters
+getDistanceBetweenPoints merchantId a b = do
```
Comment (hkmangla) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/Ride/CancelRide.hs:150]: Don't use `a` and `b` as variable name.
Rename `getDistanceBetweenPoints` properly.

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/Ride/CancelRide.hs [hunk 25]
```diff
@@ -22,35 +22,51 @@ module Domain.Action.UI.Ride.CancelRide
...
 cancelRideHandle :: ServiceHandle Flow
 cancelRideHandle =
   ServiceHandle
     { findRideById = QRide.findById,
       findById = QPerson.findById,
-      cancelRide = CInternal.cancelRideImpl
+      cancelRide = CInternal.cancelRideImpl,
+      findDriverLocationId = QDrLoc.findById,
+      findBookingByIdInReplica = Esq.runInReplica . QRB.findById,
+      pickUpDistance = getDistanceBetweenPoints
```
Comment (hkmangla) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/Ride/CancelRide.hs:69]: Need not to create handler key `getDistanceBetweenPoints` function as it being called internally only.

**Review by hkmangla — CHANGES_REQUESTED** [review]

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/Beckn/Confirm.hs [hunk 26]
```diff
@@ -326,7 +326,10 @@ cancelBooking booking mbDriver transporter = do
             rideId = (.id) <$> ride,
             source = DBCR.ByApplication,
             reasonCode = Nothing,
-            additionalInfo = Nothing
+            additionalInfo = Nothing,
```
[the remark this answers is not in the document]
Comment (abhigyanshukla97) [reply Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/Beckn/Confirm.hs:329]: pushed new changes.

File: Backend/app/rider-platform/rider-app/Main/src/Domain/Action/UI/Cancel.hs (hunk 20 above)
Comment (abhigyanshukla97) [reply Backend/app/rider-platform/rider-app/Main/src/Domain/Action/UI/Cancel.hs]: pushed new changes.

Comment (abhigyanshukla97) [reply Backend/app/rider-platform/rider-app/Main/src/Domain/Action/UI/Cancel.hs]: pushed new changes.

File: Backend/app/rider-platform/rider-app/Main/src/Domain/Action/Dashboard/Booking.hs (hunk 19 above)
Comment (abhigyanshukla97) [reply Backend/app/rider-platform/rider-app/Main/src/Domain/Action/Dashboard/Booking.hs]: pushed new changes.

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/Beckn/Init.hs [hunk 27]
```diff
@@ -105,7 +105,10 @@ cancelBooking booking transporterId = do
             rideId = Nothing,
             source = DBCR.ByApplication,
             reasonCode = Nothing,
-            additionalInfo = Nothing
+            additionalInfo = Nothing,
```
[the remark this answers is not in the document]
Comment (abhigyanshukla97) [reply Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/Beckn/Init.hs:108]: pushed new changes.

File: Backend/app/rider-platform/rider-app/Main/src/Domain/Action/UI/Cancel.hs (hunk 22 above)
Comment (abhigyanshukla97) [reply Backend/app/rider-platform/rider-app/Main/src/Domain/Action/UI/Cancel.hs]: pushed new changes.

File: Backend/app/rider-platform/rider-app/Main/src/Domain/Action/UI/Cancel.hs (hunk 23 above)
Comment (abhigyanshukla97) [reply Backend/app/rider-platform/rider-app/Main/src/Domain/Action/UI/Cancel.hs]: pushed new changes.

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/Ride/CancelRide.hs [hunk 28]
```diff
@@ -86,22 +102,28 @@ cancelRideHandler ServiceHandle {..} requestorId rideId req = withLogTag ("rideI
...
+          buildRideCancelationReason Nothing Nothing Nothing DBCR.ByMerchant ride
         DP.DRIVER -> do
           unless (authPerson.id == driverId) $ throwError NotAnExecutor
           logTagInfo "driver -> cancelRide : " ("DriverId " <> getId driverId <> ", RideId " <> getId ride.id)
-          buildRideCancelationReason (Just driverId) DBCR.ByDriver ride
+          mbLocation <- findDriverLocationId driverId
+          booking <- findBookingByIdInReplica ride.bookingId >>= fromMaybeM (BookingNotFound ride.bookingId.getId)
+          disToPickup <- forM mbLocation $ \location -> do
+            pickUpDistance booking.providerId (getCoordinates location) (getCoordinates booking.fromLocation)
+          currentDriverLocation <- forM mbLocation $ \location -> do
```
Comment (hkmangla) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/Ride/CancelRide.hs:113]: Just change to 
```
currentDriverLocation <- getCoordinates `mapM` mbLocation
```

File: Backend/app/rider-platform/rider-app/Main/src/Domain/Action/UI/Cancel.hs [hunk 29]
```diff
@@ -83,7 +88,11 @@ cancel bookingId _ req = do
     throwError $ RideInvalidStatus "Cannot cancel this ride"
   when (booking.status == SRB.NEW) $ throwError (BookingInvalidStatus "NEW")
   bppBookingId <- fromMaybeM (BookingFieldNotPresent "bppBookingId") booking.bppBookingId
-  cancellationReason <- buildBookingCancelationReason
+  ride <- Esq.runInReplica $ QR.findActiveByRBId booking.id >>= fromMaybeM (RideDoesNotExist $ "BookingId: " <> booking.id.getId)
+  res <- CallBPP.callGetDriverLocation ride.trackingUrl
```
Comment (hkmangla) [inline Backend/app/rider-platform/rider-app/Main/src/Domain/Action/UI/Cancel.hs:92]: Handle error here, if error comes from this API don't throw it, just catch it and consider driverLocation as empty

**Review by hkmangla — CHANGES_REQUESTED** [review]

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/Ride/CancelRide.hs [hunk 30]
```diff
@@ -110,5 +132,29 @@ cancelRideHandler ServiceHandle {..} requestorId rideId req = withLogTag ("rideI
             source = source,
             reasonCode = Just reasonCode,
             driverId = mbDriverId,
+            driverCancellationLocation = currentDriverLocation,
+            driverDistToPickup = disToPickup,
             ..
           }
+
+driverDistanceToPickup ::
```
Comment (shanm-16) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/Ride/CancelRide.hs:140]: why it is written seperate?

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/Ride/CancelRide.hs [hunk 31]
```diff
@@ -110,5 +132,29 @@ cancelRideHandler ServiceHandle {..} requestorId rideId req = withLogTag ("rideI
...
+    EsqDBFlow m r,
+    Metrics.CoreMetrics m,
+    Maps.HasCoordinates tripStartPos,
+    Maps.HasCoordinates tripEndPos
+  ) =>
+  Id Merchant ->
+  tripStartPos ->
+  tripEndPos ->
+  m Meters
+driverDistanceToPickup merchantId tripStartPos tripEndPos = do
```
Comment (vsaimanohar) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/Ride/CancelRide.hs:152]: change `tripStartPos` to `currentDriverLocation` and `tripEndPos` to `pickupLocation`

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/Ride/CancelRide.hs [hunk 32]
```diff
@@ -110,5 +132,29 @@ cancelRideHandler ServiceHandle {..} requestorId rideId req = withLogTag ("rideI
...
+  distRes <-
+    Maps.getDistanceForCancelRide merchantId $
+      Maps.GetDistanceReq
+        { origin = tripStartPos,
+          destination = tripEndPos,
```
Comment (vsaimanohar) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/Ride/CancelRide.hs:157]: same as above

File: Backend/app/rider-platform/rider-app/Main/src/Domain/Action/UI/Cancel.hs [hunk 33]
```diff
@@ -157,3 +168,25 @@ cancelSearch personId dcr = do
...
+  Id Merchant.Merchant ->
+  tripStartPos ->
+  tripEndPos ->
+  m Meters
+driverDistanceToPickup merchantId tripStartPos tripEndPos = do
+  distRes <-
+    Maps.getDistanceForCancelRide merchantId $
+      Maps.GetDistanceReq
+        { origin = tripStartPos,
+          destination = tripEndPos,
```
Comment (vsaimanohar) [inline Backend/app/rider-platform/rider-app/Main/src/Domain/Action/UI/Cancel.hs:189]: same as above change here too

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Storage/Tabular/BookingCancellationReason.hs (hunk 18 above)
Comment (abhigyanshukla97) [reply Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Storage/Tabular/BookingCancellationReason.hs]: pushed new changes.

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/Ride/CancelRide.hs (hunk 17 above)
Comment (abhigyanshukla97) [reply Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/Ride/CancelRide.hs]: pushed new changes.

File: Backend/app/rider-platform/rider-app/Main/src/Domain/Action/UI/Cancel.hs [hunk 34]
```diff
@@ -73,7 +78,7 @@ data CancelSearch = CancelSearch
     sendToBpp :: Bool
   }
 
-cancel :: (EncFlow m r, Esq.EsqDBReplicaFlow m r, EsqDBFlow m r, HasCacheConfig r, HedisFlow m r) => Id SRB.Booking -> (Id Person.Person, Id Merchant.Merchant) -> CancelReq -> m CancelRes
+cancel :: (EncFlow m r, Esq.EsqDBReplicaFlow m r, EsqDBFlow m r, HasCacheConfig r, HedisFlow m r, HasField "rideCfg" r Environment.RideConfig, Metrics.CoreMetrics m) => Id SRB.Booking -> (Id Person.Person, Id Merchant.Merchant) -> CancelReq -> m CancelRes
```
Comment (piyushKumar-1) [inline Backend/app/rider-platform/rider-app/Main/src/Domain/Action/UI/Cancel.hs:81]: `HasField "rideCfg" r Environment.RideConfig` this is not required

Comment (abhigyanshukla97) [reply Backend/app/rider-platform/rider-app/Main/src/Domain/Action/UI/Cancel.hs]: resolved.

**Review by piyushKumar-1 — APPROVED** [review]

**Review by hkmangla — APPROVED** [review]
